Item Rendering Extensions

Leo40Git opened the pull request:
A rewritten version of my old [item overlay customization API](https://github.com/FabricMC/fabric/pull/916).

Also adds the containing `rendering` library.

### TODO
- [x] Customizing parts of the overlay
  - [x] Count label
  - [x] Item bar
    - [x] Provide API for multiple item bars
  - [x] Cooldown overlay
- [x] Callbacks for before/after overlay rendered
- [x] Javadoc
- [x] Testmod
  - [x] Count label
  - [x] Item bars
  - [x] Cooldown overlay
  - [x] Before/after render callbacks

Leo40Git committed:
First draft of Item Rendering Extensions

Leo40Git committed:
VanillaItemBarProvider: expose as API

williambl commented:
oooh, remind me to review this later 👀

Leo40Git committed:
pretty big refactor m8

QuiltItemRenderingExtensions: add preRenderOverlay and postRenderOverlay

Rename all "Provider" interfaces to "Renderer"
CountLabelRenderer and CooldownOverlayRenderer now entirely control the rendering of their respective overlay components
Add SolidColorCooldownOverlayRenderer (like SolidColorItemBarRenderer)
Add GuiRendererHelper

Renderer interfaces are no longer client-only; rather, all their fields/methods are (so in a server environment they'd just be empty interfaces)
Re-hide VanillaItemBarRenderer as impl detail

Leo40Git committed:
GuiRendererHelper: add textured quad methods

Leo40Git committed:
VanillaCountLabelRenderer: deobfuscate constant

Leo40Git committed:
ItemRendererMixin: fix item bars being rendered WAY off target

Remove unused BufferBuilder init as well

Leo40Git committed:
liciense

oliviathevampire commented:
This looks pretty good! but waiting for other people to review it, you need to update to the latest changes, and I suggets adding a test mod.

oliviathevampire commented on library/item/item_rendering/build.gradle:
```diff
@@ -0,0 +1,19 @@
+		}
+	}
+	injectedInterface("net/minecraft/class_1792") {
+		values = ["org/quiltmc/qsl/item/rendering/client/api/QuiltItemRenderingExtensions"]
```
this should be ``org/quiltmc/qsl/item/rendering/api/client/QuiltItemRenderingExtensions`` instead of ``org/quiltmc/qsl/item/rendering/client/api/QuiltItemRenderingExtensions``

Leo40Git committed:
Fix build

Leo40Git committed:
Merge remote-tracking branch 'upstream/1.19' into feature/item-rendering

Leo40Git committed:
Allow cancelling overlay rendering

Leo40Git committed:
Expose VanillaCountLabel/CooldownOverlayRenderers as API

0xJoeMama commented:
Can we get BuiltinItemRendererRegistry as well with this PR? Or do you think it's out of scope?

Leo40Git committed:
Improve SolidColor classes a bit

Leo40Git committed:
oops, add default return value to preRenderOverlay

Leo40Git committed:
Fix all items having durability bar

Fix multiple item bars starting 2 pixels too low

Leo40Git committed:
T E S T M O D

todo: pre/postRenderOverlay tests

Leo40Git commented:
> Can we get BuiltinItemRendererRegistry as well with this PR? Or do you think it's out of scope?

I'll do that in a separate PR after this one gets merged.

williambl commented on library/item/item_rendering/src/main/java/org/quiltmc/qsl/item/rendering/mixin/client/ItemRendererMixin.java:
```diff
@@ -0,0 +1,99 @@
+	 * @author QuiltMC
+	 * @reason Completely refactoring overlay rendering logic to use {@link QuiltItemRenderingExtensions}
+	 */
+	@Overwrite
```
This overwrite is going to cause a lot of incompatibilities with fabric mods (
![image](https://user-images.githubusercontent.com/5115825/179732138-18099265-50be-4627-a7be-6dfbc80529c9.png)
) - is there any other way we can do it?

Leo40Git replied:
I made it an `@Overwrite` simply because I couldn't be bothered to redo my old many-parts mixin at the time. I'll try to split it up into `@Inject`s and `@Redirect`s.

Leo40Git replied:
...also, that third one is 100% copied from my old PR! It has the same comments as well.

williambl reviewed (changes requested):
I am super happy with the API itself. Tests are great too. I only see one problem, the overwrite, since some fabric mods will break with it.

OroArmor commented:
You might hate this, but i think this would fit better under the `rendering` library as an `item` module. Since this entire module is client side and rendering related, i think it fits better there.

0xJoeMama commented:
Agreed ^^^

I only just realised this was under the item library.

Leo40Git committed:
Move to rendering library

Leo40Git committed:
Extend GuiRendererHelper

Leo40Git committed:
Add pre/postRenderOverlay tests to testmod

THE RETURN OF THE HENRY REFERENCE

Leo40Git commented:
IT HATH BEEN MOVED

Leo40Git committed:
license, as usual

Leo40Git committed:
Make Item Rendering Extensions compatible with all mods, ever

(Okay, slight exaggeration.)
Instead of simply @Overwrite-ing renderGuiItemOverlay, this module will now use the alternate "customized" rendering path ONLY if the provided item has a non-vanilla renderer set.
If the item has no custom renderers, the Vanilla rendering path is taken - preserving any injections/redirections/whatever a mod might inject into the method to customize its items' overlay rendering.

Additionally, the Vanilla rendering path has been slightly optimized by avoiding allocating a MatrixStack instance every time it's called (which is... a LOT of times).

QuiltItemRenderingExtensions: getItemRenderers() now returns null by default
ItemBarRenderer: remove getDefaultRenderers()

Leo40Git committed:
Javadoc it up

Leo40Git committed:
GuiRendererHelper: javadoc

Leo40Git committed:
ItemRendererMixin: move "stack is empty" check up

williambl reviewed (approved):
Beautiful. Maybe give it a package-info with a high-level overview javadoc? Otherwise, perfect.

Leo40Git committed:
Add package-info.java

Leo40Git committed:
ItemRendererMixin: avoid additional ThreadLocal.get calls

fix postRenderOverlay being called for empty ItemStacks

Leo40Git committed:
Javadoc polish

0xJoeMama commented on library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/GuiRendererHelper.java:
```diff
@@ -0,0 +1,306 @@
+/**
+ * Contains various helper methods for rendering GUI elements.
+ */
+public interface GuiRendererHelper {
```
To be frank, I think this interface should be split into 2 others. One that holds the static methods, which are general utility methods(you could debate this could be moved to another module as well), and another one with the non-static ones.

Leo40Git replied:
Honestly I should just make this a proper utility class

0xJoeMama commented on library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/CooldownOverlayRenderer.java:
```diff
@@ -0,0 +1,47 @@
+/**
+ * Renders a cooldown overlay.
+ */
+@FunctionalInterface
```
I suppose you could annotate the whole interface with Environment, instead of doing so on its members.

Leo40Git replied:
This was purposefully done to prevent any possible "class does not exist" shenanigans on dedicated servers - the interface exists on all sides, but is empty outside of clients.

0xJoeMama replied:
~~Wouldn't this be the perfect use case for `@EnvironmentalInterface`?~~
Forget this, seems I misunderstood was EnvironmentInterface does.

0xJoeMama commented on library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java:
```diff
@@ -0,0 +1,167 @@
+
+@Mixin(ItemRenderer.class)
+public abstract class ItemRendererMixin {
+	@Unique private final ThreadLocal<MatrixStack> quilt$matrices = ThreadLocal.withInitial(MatrixStack::new);
```
Do we really need to use ThreadLocal here? I don't think this is accessed by any other thread, other than the Render Thread.

OroArmor replied:
Its good practice. Other mods might also render twice like Replay mod (Not sure)

0xJoeMama commented on library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java:
```diff
@@ -0,0 +1,167 @@
+	}
+
+	@Unique
+	private boolean isItemOverlayRenderingCustomized(Item item) {
```
Does adding `@Unique` make sure that the method never collides? 
Does it salt the method or something? If it doesn't I suppose you could prefix it with quilt$ to ensure nothing weird happens.

Leo40Git replied:
`@Unique` with private methods 100% makes the method never collide, AFAIK.

0xJoeMama replied:
![image](https://user-images.githubusercontent.com/46004440/180822157-fff6a287-22ef-4aab-91b2-a699a20cfcda.png)
Yeap javadocs confirms that to be the case.

0xJoeMama commented on library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java:
```diff
@@ -0,0 +1,167 @@
+	}
+
+	@Unique
+	private boolean isItemOverlayRenderingCustomized(Item item) {
```
I also think this should be in a separate class. Might be useful for people.

0xJoeMama commented on library/rendering/item_rendering/src/testmod/java/org/quiltmc/qsl/rendering/item/test/client/cooldown/FlashingCooldownOverlayRenderer.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * Shows a full red, flashing overlay if there's more than 80% of the cooldown still remaining.
+ */
+public class FlashingCooldownOverlayRenderer extends VanillaCooldownOverlayRenderer {
```
You might want to add an Environment annotation here.

0xJoeMama commented on library/rendering/item_rendering/src/testmod/java/org/quiltmc/qsl/rendering/item/test/client/cooldown/HiddenCooldownOverlayRenderer.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Hides the cooldown overlay, even if there is a cooldown, as long as it has more than 20% remaining.
+ */
+public class HiddenCooldownOverlayRenderer extends VanillaCooldownOverlayRenderer {
```
Same here.

0xJoeMama commented on library/rendering/item_rendering/src/testmod/java/org/quiltmc/qsl/rendering/item/test/client/ItemDecorations.java:
```diff
@@ -0,0 +1,68 @@
+public class ItemDecorations {
+	private ItemDecorations() { }
+
+	public static final Identifier STACK_BORDER_TEXTURE = ItemRenderingTestmod.id("textures/gui/border.png");
```
Hm. Are these textures stitched into the item texture Atlas? If not, this may cause some performance issues if a lot of mods follow this trend.

Leo40Git replied:
They are not, I have no idea how to do that

OroArmor replied:
Texture atlases are extremely complex, and the GUI texture usually used is not atlased. This would be quite an invasive change to Minecraft if this were to be fixed. Not sure the best way to fix it though.

0xJoeMama commented on library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/GuiRendererHelper.java:
```diff
@@ -0,0 +1,306 @@
+
+/**
+ * Contains various helper methods for rendering GUI elements.
+ */
```
`@Environment`

0xJoeMama commented on library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/VanillaCountLabelRenderer.java:
```diff
@@ -0,0 +1,83 @@
+
+		matrices.push();
+		matrices.translate(0, 0, zOffset + 200);
+		// TODO figure out if we can NOT render immediately here
```
I suppose this TODO means you haven't found a way to avoid Immediate right? That's a bit concerning.

0xJoeMama replied:
I guess vanilla uses Immediate here as well right? I'll have to go look, but I think that's the case.

Leo40Git replied:
Yeah, this is copied from vanilla.

0xJoeMama reviewed (commented):
General Note: Mixin handler methods, are salted at runtime, therefore adding quilt$ in the beginning doesn't really offer a lot in terms of compatibility, while also making the code a bit more annoying to read. Correct me if I am wrong.

I'd also like to note that this review doesn't offer much, in terms of getting the PR approved. It's just some comments I could scratch up.

Leo40Git committed:
Convert GuiRendererHelper to proper util class

Leo40Git committed:
Expose ItemRendererMixin.isItemOverlayRenderingCustomized

New name is QuiltItemRendering.areOverlayComponentsCustomized

Leo40Git committed:
testmod: add missing Environment annotations

Leo40Git committed:
what: how did these turn into spaces??

OroArmor commented on library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/GuiRendererHelper.java:
```diff
@@ -0,0 +1,183 @@
+		buffer.vertex(mat, x, y + height, 0.0f).color(r, g, b, a).next();
+		buffer.vertex(mat, x + width, y + height, 0.0f).color(r, g, b, a).next();
+		buffer.vertex(mat, x + width, y, 0.0f).color(r, g, b, a).next();
+		BufferRenderer.drawWithShader(buffer.end());
```
I can hear jellysquid screaming a thousand screams as someone loops this method and submits 10k draw calls, bringing the FPS down to 0. I would remove the draw call and only set the shader if the current shader is different and draw the buffer as it is. Otherwise this is *extremely* inefficient code and bad for rendering.

Leo40Git replied:
Yeah, I _really_ need to figure out how to make this not horrible (AKA sharing a single `BufferBuilder` and drawing it at the end) while also not making it break should someone set a new shader and not set it back.

OroArmor replied:
I would look at the `DrawableHelper` class. I do think it covers most of these methods.

OroArmor commented on library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/QuiltItemRenderingExtensions.java:
```diff
@@ -0,0 +1,85 @@
+	 * @return array of item bar renderers, or {@code null} to use the vanilla renderer only
+	 */
+	@Environment(EnvType.CLIENT)
+	default ItemBarRenderer @Nullable [] getItemBarRenderers() {
```
This `@Nullable` annotation is in a weird place. i would do `@Nullable ItemBarRenderer[]`

Leo40Git replied:
`@Nullable ItemBarRenderer[]` -> array itself has no annotation, elements are `@Nullable`
`ItemBarRenderer @Nullable []` -> array itself is `@Nullable`, elements have no annotations (though they're treated as `@NotNull` in the impl)

OroArmor replied:
Ah ok. Looks weird to me but makes sense.

OroArmor commented on library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java:
```diff
@@ -0,0 +1,158 @@
+	}
+
+	@Unique
+	private void renderCustomGuiItemOverlay(MatrixStack matrices,
+										   TextRenderer renderer, ItemStack stack, @Nullable String countLabel) {
```
The line wrapping and formatting on these are a bit weird. I think making your line length longer would fix it. I can't remember if QSL has a limit

OroArmor reviewed (changes requested):
Small nitpicks. I wonder if an accesswidener for the model predicate provider registry could be made in this PR as well.

Leo40Git committed:
Merge remote-tracking branch 'upstream/1.19' into feature/item-rendering

# Conflicts:
#	settings.gradle

Leo40Git committed:
draft 1 of Operation: Desuckify Overlay Rendering Performance (DORP)

QuadHelper will be passed to all item overlay rendering methods, essentially acting as a quad draw batch manager
getBufferBuilder, reinitialize and flush are low-level methods to allow devs to do a funni
beginQuad and beginTexturedQuad are high-level methods intended for most devs

Leo40Git committed:
Revert "ItemRendererMixin: avoid additional ThreadLocal.get calls"

This reverts commit 2fe216e5324599671fc95c91a8e00c13015c5ea9.

Leo40Git committed:
rename QuadHelper to QuadBatchManager

add missing license headers

Leo40Git committed:
QuadBatchManagerImpl: untrack state when ending current batch

Leo40Git committed:
pass QuadBatchManager to all rendering methods

Leo40Git committed:
QuadBatchManagerImpl: more rendering state management

depth test is always disabled by default

Leo40Git committed:
replace GuiRendererHelper with QuadBuilder

update all rendering to use QuadBatchManager + QuadBuilder

Leo40Git committed:
add missing licenese headers

Leo40Git committed:
VanillaCountLabelRenderer: properly grab BufferBuilder

Leo40Git commented:
I have committed an initial version of the quad batching API (currently untested since I can't run the testmod for some reason).

Leo40Git committed:
Mark item_rendering module as client-only

Leo40Git committed:
ItemRendererMixin: fix some indents

Leo40Git committed:
Merge branch '1.19' into feature/item-rendering

Leo40Git commented:
Closing because _man_ I made this overcomplicated as hell. Will bring back later (possibly after extracting the quad batching stuff into its own PR)?

maximumpower55 commented:
Feels like QSL shouldn't be doing this invasive rendering optimizations, quad batching would fit better in its own rendering optimization mod Imo

Leo40Git commented:
This optimization only applies to items customized via this API - Vanilla items (and items from mods that do _not_ use this API) are unaffected.

maximumpower55 commented:
Ah👍
